jetty-cdi/src/main/java/org/eclipse/jetty/cdi/CdiSpiDecorator.java

```diff
@@ -92,6 +95,35 @@ public CdiSpiDecorator(ServletContextHandler context) throws UnsupportedOperatio
         }
     }
 
+    /**
+     * Test if a class can be decorated.
+     * @implNote The default implementation calls {@link #isKnownUndecoratable(String) }
+     * on the class and all it's super classes.
+     * @param clazz The class to check
+     * @return True if the class and all it's super classes can be decorated
+     */
+    protected boolean isDecoratable(Class<?> clazz)
+    {
+        if (Object.class == clazz)
+            return true;
+        if (isKnownUndecoratable(clazz.getName()))
+            return false;
+        return isDecoratable(clazz.getSuperclass());
+    }
+
+    /**
+     * Test if a specific class name is known to not be decoratable.
+     * @implNote default implementation checks for well known classes that are used to
+     * setup CDI itself, and thus cannot themselves be decorated.
+     * @see #isDecoratable(Class)
+     * @param className The name of the class to check
+     * @return True if the class is known not to be decoratable
+     */
+    protected boolean isKnownUndecoratable(String className)
+    {
+        return "org.jboss.weld.environment.servlet.Listener".equals(className);
```

Comment:
This complex little dance is because it is too easy to install `CdiSpiDecorator` and then try to init Weld with it's listener, which will fail because `CdiSpiDecorator` will try to decorate it, but CDI is not enabled yet.     This soft dependency  breaks that chicken-egg problem.